### ISSUE TITLE
[RFC] Compilers, cross-compilers, dependencies

### ISSUE THREAD

**Issue by mosteo** [alire-project/alire#728]
This is just a bit of a musing, to see it can go anywhere. In the context of #723 and crates requiring different compilers (native, cross-compilers), I wonder if an old feature that is lurking unused could come in handy, which is the "equivalent" crates (the "provides" alias of a release).

So, let's assume a dependency A requires a modern GNAT version. This is currently expressed for example as `gnat = ">=10"`. This is a utility library that doesn't care about the target.

Now, there's our project B which we know requires a cross-compiler `gnat-cross-X`. Obviously, both crates should be compiled with the same compiler but they do not refer to the same dependency, so currently that's a no-no.

The equivalent crate feature comes in now. This was inspired by virtual `apt` packages ([here](https://www.debian.org/doc/debian-policy/ch-binary.html#s-virtual-pkg) and [here](https://www.debian.org/doc/debian-policy/ch-relationships.html#s-virtual)). Let's say that we redefine our compiler crates this way (let's forget what `gnat` currently means):

- `gnat`: this is the "virtual" crate: any other crate may indicate that it provides this dependency. This is not an actual crate in the index.
- `gnat-native`: a native compiler (from binares, system packages...)
- `gnat-cross-X`: an example cross compiler.

In the crate definitions of `gnat-native` and `gnat-cross`, they should indicate that they fulfill the `gnat` dependency (`provides = "gnat"`). During resolution, the solver accepts any crate that provides the virtual one. In this example, due to B depending on `gnat-cross`, and A depending on any gnat, `gnat-cross` will be used.

Now, for the case where crates only specify any `gnat`, if we detect a installed `gnat` (via #723, or the system compiler, or manually), that one should take precedence over other alternatives (this is unimplemented).

Finally, if a crate depends on `gnat-native` and another one on `gnat-cross`, or two different cross-compilers, these are obviously incompatible. (I wonder if "or" dependencies (also lurking unused) might also serve for crates that can target several boards, but that's a separate topic.)

In short, the `provides` field creates an alias that unifies a family of crates. Dependents can refer either to any from the family via the alias, or concretely to an actual crate.

Ping @Fabien-Chouteau because I botched the initial issue creation...

**Comment by Lucretia** [comment]
I just mentioned on the [forum](https://forum.ada-lang.io/t/alire-should-provide-cross-compilers/1053/2) that we should have cross compilers available for us to test builds with.

**Comment by Fabien-Chouteau** [comment]
@Lucretia, we don't have the setup and resources to make more compilers, unfortunately.
That's something we will work on in the next months.

**Comment by ThyMYthOS** [comment]
Regarding cross compilation, the compiler itself (same is true for gnatformat, gnatprove, ...) should not be a "normal" dependency specified by `depends-on`, but instead a dependency for the build platform.

For cross compilation we typically distinguish three platforms:

- **Build Platform:** This is the platform on which the compilation tools are executed.
- **Host Platform:** This is the platform on which the code will eventually run.
- **Target Platform:** If this is a compiler, this is the platform that the compiler will generate code for.

So the regular `depends-on` would only specify host platform dependencies, while something like `build-depends-on` might specify dependencies for the build platform. The target platform is part of the name of the toolchain.

**Comment by mosteo** [comment]
This feature discussed here was actually implemented long ago with the `provides` keyword, so I'm closing this one.

I've opened #2067 to continue discussion on the separate sets of dependencies. @ThyMYthOS, I answer you there.
